logtail/logtail_test.go

```diff
@@ -23,28 +27,213 @@ func TestFastShutdown(t *testing.T) {
 	l := NewLogger(Config{
 		BaseURL: testServ.URL,
 	}, t.Logf)
-	l.Shutdown(ctx)
+	err := l.Shutdown(ctx)
+	if err != nil {
+		t.Error(err)
+	}
 }
 
-func TestUploadMessages(t *testing.T) {
-	ctx, cancel := context.WithCancel(context.Background())
-	uploads := 0
-	testServ := httptest.NewServer(http.HandlerFunc(
+// maximum number of times a test will call l.Write()
+const logLines = 3
+
+type LogtailTestServer struct {
+	srv      *httptest.Server // Log server
+	uploaded chan []byte
+}
+
+func NewLogtailTestServer(t *testing.T) *LogtailTestServer {
+	ts := LogtailTestServer{}
+
+	// max channel backlog = 1 "started" + #logLines x "log line" + 1 "closed"
+	ts.uploaded = make(chan []byte, 2+logLines)
+
+	ts.srv = httptest.NewServer(http.HandlerFunc(
 		func(w http.ResponseWriter, r *http.Request) {
-			w.Header().Set("Content-Type", "application/json; charset=utf-8")
-			uploads += 1
+			body, err := ioutil.ReadAll(r.Body)
+			if err != nil {
+				t.Error("failed to read HTTP request")
+			}
+			ts.uploaded <- body
 		}))
-	defer testServ.Close()
 
-	l := NewLogger(Config{BaseURL: testServ.URL}, t.Logf)
-	for i := 1; i < 10; i++ {
+	t.Cleanup(ts.srv.Close)
+	return &ts
+}
+
+func (ts *LogtailTestServer) HandleStartLoggingMessage() {
+	// There is always an initial "logtail started" message
+	body := <-ts.uploaded
+	if !strings.Contains(string(body), "started") {
+		panic("unknown start logging statement:" + string(body))
+	}
+}
+
+func TestDrainPendingMessages(t *testing.T) {
+	ts := NewLogtailTestServer(t)
+	l := NewLogger(Config{BaseURL: ts.srv.URL}, t.Logf)
+	ts.HandleStartLoggingMessage()
+
+	for i := 0; i < logLines; i++ {
 		l.Write([]byte("log line"))
 	}
 
-	l.Shutdown(ctx)
-	cancel()
-	if uploads == 0 {
-		t.Error("no log uploads")
+	// all of the "log line" messages usually arrive at once, but poll if needed.
+	body := ""
+	for i := 0; i <= logLines; i++ {
+		body += string(<-ts.uploaded)
+		cnt := strings.Count(body, "log line")
+		if cnt == logLines {
+			break
+		}
+	}
+
+	err := l.Shutdown(context.Background())
+	if err != nil {
+		t.Error(err)
+	}
+
+	if strings.Count(body, "log line") != logLines {
+		t.Errorf("got %q; want: %d copies of %q", body, logLines, "log line")
+	}
+}
+
+func TestEncodeAndUploadMessages(t *testing.T) {
+	ts := NewLogtailTestServer(t)
+	l := NewLogger(Config{BaseURL: ts.srv.URL}, t.Logf)
+	ts.HandleStartLoggingMessage()
+
+	tests := []struct {
+		name string
+		log  string
+		want string
+	}{
+		{
+			"plain text",
+			"log line",
+			"log line",
+		},
+		{
+			"simple JSON",
+			`{"text": "log line"}`,
+			"log line",
+		},
+	}
+
+	for _, tt := range tests {
+		io.WriteString(l, tt.log)
+		body := <-ts.uploaded
+
+		data := make(map[string]interface{})
+		err := json.Unmarshal(body, &data)
+		if err != nil {
+			t.Error(err)
+		}
+
+		got, ok := data["text"]
+		if ok {
```

Comment:
You could simplify this a bit. If there's no "text" entry in data, then `got` will be the zero value (`""`). So you could omit the comma-ok and do:

```go
got := data["text"]
if got != tt.want {
  t.Errorf(...)
}
```

Reply:
Done